dot/sync/disjoint_block_set.go

```diff
@@ -58,50 +85,79 @@ type pendingBlock struct {
 //
 // if the block is complete, we may not know of its parent.
 type disjointBlockSet struct {
-	// TODO: put a limit on the size of this set
 	sync.RWMutex
+	limit int
+
+	// map of block hash -> block data
 	blocks map[common.Hash]*pendingBlock
+
+	// map of parent hash -> child hashes
+	parentToChildren map[common.Hash]map[common.Hash]struct{}
 }
 
-func newDisjointBlockSet() *disjointBlockSet {
+func newDisjointBlockSet(limit int) *disjointBlockSet {
 	return &disjointBlockSet{
-		blocks: make(map[common.Hash]*pendingBlock),
+		blocks:           make(map[common.Hash]*pendingBlock),
+		parentToChildren: make(map[common.Hash]map[common.Hash]struct{}),
+		limit:            limit,
 	}
 }
 
-func (s *disjointBlockSet) addHashAndNumber(hash common.Hash, number *big.Int) {
+func (s *disjointBlockSet) addToParentMap(parent, child common.Hash) {
+	children, has := s.parentToChildren[parent]
+	if !has {
+		children = make(map[common.Hash]struct{})
+		s.parentToChildren[parent] = children
```

Comment:
do we need mutex on this?

Reply:
it's only used within other `disjointBlockState` functions that already have the mutex